pallets/utxo/src/lib.rs

```diff
@@ -478,38 +519,54 @@ pub mod pallet {
             );
         }
 
-        let full_inputs: Vec<(crate::TokenID, TransactionOutputFor<T>)> = tx
+        let full_inputs: Vec<(TokenId, TransactionOutputFor<T>)> = tx
             .inputs
             .iter()
             .filter_map(|input| <UtxoStore<T>>::get(&input.outpoint))
-            .map(|output| (OutputHeaderData::new(output.header).token_id(), output))
+            .filter_map(|output| {
+                if let TxHeaderAndExtraData::NormalTx { id } = &output.header {
+                    Some((*id, output))
+                } else {
+                    None
+                }
+            })
             .collect();
 
-        let input_vec: Vec<(crate::TokenID, Value)> =
+        let input_vec: Vec<(TokenId, Value)> =
             full_inputs.iter().map(|output| (output.0, output.1.value)).collect();
 
-        let out_vec: Vec<(crate::TokenID, Value)> = tx
+        let out_vec: Vec<(TokenId, Value)> = tx
             .outputs
             .iter()
-            .map(|output| {
-                (
-                    OutputHeaderData::new(output.header).token_id(),
-                    output.value,
-                )
+            .filter_map(|output| {
+                if let TxHeaderAndExtraData::NormalTx { id } = &output.header {
+                    Some((*id, output.value))
+                } else {
+                    None
+                }
             })
             .collect();
 
         // Check for token creation
         let tokens_list = <TokenList<T>>::get();
```

Comment:
We're returning a vector here, and I think this is very inefficient... at least a BTreeSet should be returned out of this to have log(N) complexity when searching for tokens at worst instead of O(N). Or alternatively a HashSet to make it as close to O(1) as possible. Searching in a vector linearly seems like the worst way to do this, and can even open the door to DOS attacks.

I'm assuming here there are no duplicates. If there are no duplicates, you should convert the return value to a set or hash set before searching in it like you do below. Otherwise, we should have a discussion on how to do this.

Even better, you should be able to search for a token in the storage directly... why retrieve all tokens and search in them?!

Reply:
@TheQuantumPhysicist Sorry but your answer is completely not related to this code, probably you have selected the wrong line? 

As an input we have Tx.Outputs that is equal `pub(crate) outputs: Vec<TransactionOutput<AccountId>>,`.
I should calculate all MLT that we use as a fee for transactions. Fee calculated like MLT from all outputs minus MLT from all inputs. 

We anyway will be going through all the `outputs` never mind what data container we will use, BTreeSet \ BTreeMap \ etc all of these require mapping the whole Vec.

What we can do `fn validate_transaction`? Reduce the number of Vectors and probably refactor it but it doesn't relate to this PR. @muursh

Reply:
As @TheQuantumPhysicist mentioned in DM, this comment related to change the type of the `TokenList`. I'll do that